### PR TITLE
[go_router] Use `leak_tracker_flutter_testing` 

### PR DESCRIPTION
**Pull request by ValentinVignal**
Activate `leak_tracker_flutter_testing` in the tests to detect memory leak issues.

cc @polina-c 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [ ] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by ValentinVignal** [comment]
@polina-c Is there an issue to link this PR to?

**Comment by ValentinVignal** [comment]
```
Because go_router depends on flutter_test from sdk which depends on leak_tracker_flutter_testing 3.0.3, leak_tracker_flutter_testing 3.0.3 is required.
So, because go_router depends on leak_tracker_flutter_testing ^2.0.1, version solving failed.
```

But `leak_tracker_flutter_testing ^2.0.1` is the version available on stabe. @polina-c how can I make it work?

**Comment by polina-c** [comment]
> ```
> Because go_router depends on flutter_test from sdk which depends on leak_tracker_flutter_testing 3.0.3, leak_tracker_flutter_testing 3.0.3 is required.
> So, because go_router depends on leak_tracker_flutter_testing ^2.0.1, version solving failed.
> ```
> 
> But `leak_tracker_flutter_testing ^2.0.1` is the version available on stabe. @polina-c how can I make it work?

Will investigate. Thank you.

**Comment by polina-c** [comment]
> > ```
> > Because go_router depends on flutter_test from sdk which depends on leak_tracker_flutter_testing 3.0.3, leak_tracker_flutter_testing 3.0.3 is required.
> > So, because go_router depends on leak_tracker_flutter_testing ^2.0.1, version solving failed.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But `leak_tracker_flutter_testing ^2.0.1` is the version available on stabe. @polina-c how can I make it work?
> 
> Will investigate. Thank you. Converted the PR to draft for now.

Okay, it should be `leak_tracker_flutter_testing: any` in pubspec.yaml to peck up leak tracker from sdk. Thanks for raising this. I will update documentation.

**Comment by ValentinVignal** [comment]
Great thank you! I changed it in [chore: Use any version for leak_tracker_flutter_testing](https://github.com/flutter/packages/pull/6210/commits/f1804220c9cc50b0f63b2d861dfc7b0ea8b5f56a)

**Comment by polina-c** [comment]
I see tests are failing. And they are failing because of leaks. And many of them are originated by flutter itself.

For example, leaking "_NotAnnounced" was cleaned up this week in Flutter master. 

It seems packages that use `stable` should not be early adopters of leak tracker. So, let's shelve this PR till leak_tracker is released and recommended for stable.

Thank you for the experiment. It helped:
1. To understand go_router has leaks.
2. To document that version should be `any`.

Converting this to draft.

**Comment by ValentinVignal** [comment]
@chunhtai what do you want me to do? Maybe I can remove `leak_tracker_flutter_testing` from the `pubspec.yaml` and revert the `flutter_test_config.dart` but keep the changes to prepare for when `leak_tracker_flutter_testing` can be used ?

**Comment by polina-c** [comment]
> @chunhtai what do you want me to do? Maybe I can remove `leak_tracker_flutter_testing` from the `pubspec.yaml` and revert the `flutter_test_config.dart` but keep the changes to prepare for when `leak_tracker_flutter_testing` can be used ?

Yes, I like it: to merge fixes, while postponing regression testing. @chunhtai, how does it sound for you?

**Comment by polina-c** [comment]
> > @chunhtai what do you want me to do? Maybe I can remove `leak_tracker_flutter_testing` from the `pubspec.yaml` and revert the `flutter_test_config.dart` but keep the changes to prepare for when `leak_tracker_flutter_testing` can be used ?
> 
> Yes, I like it: to merge fixes, while postponing regression testing. @chunhtai, how does it sound for you?

[chunhtai](https://github.com/chunhtai) is unavailable at the moment
I think we can proceed with merging the change.

**Comment by ValentinVignal** [comment]
I removed the `leak_tracker_flutter_testing` from the `pubspec.yaml` and revert the `flutter_test_config.dart` in [test: Disable leak tracking](https://github.com/flutter/packages/pull/6210/commits/c8d36467b8d00b8d7f789de5b60519c364449a77)

**Review by polina-c — APPROVED** [review]

**Comment by polina-c** [comment]
'Linux analyze_legacy N-2' is failing because HeroController did not have method dispose yet. 

@johnpryan, is this legacy job fixed forever or we can upgrade it for newer versions of flutter?

**Review by johnpryan — APPROVED** [review]
LGTM

**Comment by polina-c** [comment]
> 'Linux analyze_legacy N-2' is failing because HeroController did not have method dispose yet.
> 
> @johnpryan, is this legacy job fixed forever or we can upgrade it for newer versions of flutter?

More information: https://github.com/flutter/packages/blob/main/.ci/legacy_project/README.md

@ValentinVignal, invoked dispose does not exist for this job, that is on Flutter 2.0.6. What is minimal version of Flutter that has this method?
We either need to increase oldest supported version of flutter, or not to add this missing dispose (that may cause memory leaks)

@stuartmorgan, are you right person to help here? Is it ok to stop supporting 2.0.6 and increase minimal supported version a little?

**Comment by stuartmorgan** [comment]
> @stuartmorgan, are you right person to help here? Is it ok to stop supporting 2.0.6 and increase minimal supported version a little?

The earliest version of Flutter that any package in this repo is allowed to support (enforced by CI) is 3.13, so I don't really understand the question.

If N-2 is failing, then the PR doesn't work in 3.13.9. It's fine to drop 3.13 [per the wiki](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#when-to-update-the-required-version), but that's unrelated to 2.0.6.

**Comment by polina-c** [comment]
> https://github.com/flutter/packages/blob/main/.ci/legacy_project/README.md

Thanks.

I took 2.0.6 from https://github.com/flutter/packages/blob/main/.ci/legacy_project/README.md, but now i see it is just very first supported version.

**Comment by ValentinVignal** [comment]
I'm a bit unsure of what I should do, should I stop using `HeroController.dispose` or should I make a change to the pipeline?

**Comment by polina-c** [comment]
> I'm a bit unsure of what I should do, should I stop using `HeroController.dispose` or should I make a change to the pipeline?

To choose between these two options we need to find out what pipeline change is needed. Then we will decide if this change makes sense or if it is better not to use dispose for now and instead create tech debt issue and leave TODO to dispose when all supported versions of flutter have the dispose. Does it help?

**Comment by ValentinVignal** [comment]
@polina-c After looking at it, it is because `HeroControlle.dispose` was released in flutter `3.16.x` 
- https://github.com/flutter/flutter/pull/133951
- https://docs.flutter.dev/release/release-notes/release-notes-3.16.0
So I had to increase the minimum flutter version of flutter to `^3.16.0`

Is that okay? Or is it too soon?

File: packages/go_router/example/pubspec.yaml [hunk 1]
```diff
@@ -5,7 +5,7 @@ publish_to: none
 
 environment:
   sdk: ^3.1.0
```
Comment (stuartmorgan) [inline packages/go_router/example/pubspec.yaml:7]: You'll need to change this to match (3.2.0), which is why CI is failing.

File: packages/go_router/CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,7 @@
-## NEXT
+## 13.2.1
 
-* Updates minimum supported SDK version to Flutter 3.13/Dart 3.1.
+- Updates minimum supported SDK version to Flutter 3.13/Dart 3.1.
```
[the remark this answers is not in the document]
Comment (ValentinVignal) [reply packages/go_router/CHANGELOG.md]: Oops sorry about that.

I updated it in
[doc: Update the changelog to specify the new flutter and dart versions](https://github.com/flutter/packages/pull/6210/commits/081b18374e936cab93c9e4ba691a2388c07633ba)

**Pull request merged**